File: ja2/src/disk/DiskBytes.java
package disk;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.concurrent.atomic.AtomicBoolean;

/*
 * Created on Sep 16, 2007
 */
public class DiskBytes
{
	private static final int EOF = -1;
	private static final int BYTES_PER_TRACK = 0x1A00;
	private static final int TRACKS_PER_DISK = 0x23;

	private byte[][] bytes;

	private volatile File file;
	private volatile String fileName;
	private final AtomicBoolean writable = new AtomicBoolean();
	private final AtomicBoolean loaded = new AtomicBoolean();
	private int byt;
	private int waitTimes;

	public DiskBytes()
	{
		unload();
	}

	public void load(File f) throws IOException, InvalidDiskImage
	{
		if (isLoaded())
		{
			unload();
		}
		f = f.getCanonicalFile();
		if (!f.exists())
		{
			throw new FileNotFoundException("Can't find file: "+f.getPath());
		}
		if (f.length() != TRACKS_PER_DISK*BYTES_PER_TRACK)
		{
			throw new InvalidDiskImage(f.getName());
		}
		byte[][] rb = new byte[TRACKS_PER_DISK][BYTES_PER_TRACK];
		final InputStream disk = new BufferedInputStream(new FileInputStream(f));
		int itrack = 0;
		for (int b1 = disk.read(); b1 != EOF && itrack < TRACKS_PER_DISK; b1 = disk.read())
		{
			if (b1 < 0x96)
			{
				throw new InvalidDiskImage(f.getName());
			}
			rb[itrack][this.byt] = (byte)b1;
			if (nextByte())
			{
				++itrack;
			}
		}
		disk.close();

		this.file = f;
		this.fileName = this.file.getName();
		this.bytes = rb;
		synchronized (this.writable)
		{
			this.writable.set(f.canWrite());
		}
		synchronized (this.loaded)
		{
			this.loaded.set(true);
			this.loaded.notifyAll();
		}
	}

	public String getFileName()
	{
		return this.fileName;
	}

	public boolean isLoaded()
	{
		synchronized (this.loaded)
		{
			return this.loaded.get();
		}
	}
	public void save() throws IOException
	{
		if (isWriteProtected() || !isLoaded())
		{
			return;
		}
		final OutputStream out = new BufferedOutputStream(new FileOutputStream(this.file),TRACKS_PER_DISK*BYTES_PER_TRACK);
		for (int itrack = 0; itrack < TRACKS_PER_DISK; ++itrack)
		{
			for (int ibyte = 0; ibyte < BYTES_PER_TRACK; ++ibyte)
			{
				out.write(this.bytes[itrack][ibyte]);
			}
		}
		out.flush();
		out.close();
	}

	public void unload()
	{
		this.byt = 0;
		synchronized (this.writable)
		{
			this.writable.set(true);
		}
		synchronized (this.loaded)
		{
			this.loaded.set(false);
			this.loaded.notifyAll();
		}
		this.file = null;
		this.fileName = "";
	}

	public byte get(final int track)
	{
		if (track < 0 || TRACKS_PER_DISK <= track)
		{
			throw new IllegalStateException();
		}
		synchronized (this.loaded)
		{
			if (!this.loaded.get())
			{
				++this.waitTimes;
				if (this.waitTimes >= 0x100)
				{
					this.waitTimes = 0;
					try
					{
						this.loaded.wait(20);
					}
					catch (InterruptedException e)
					{
						e.printStackTrace();
						Thread.currentThread().interrupt();
					}
				}
			}
			if (!this.loaded.get())
			{
				return -1;
			}
		}
		final byte ret = this.bytes[track][this.byt];
		nextByte();
		return ret;
	}

	void put(final int track, final byte value)
	{
		if (track < 0 || TRACKS_PER_DISK <= track)
		{
			throw new IllegalStateException();
		}
		if (isWriteProtected() || !isLoaded())
		{
			return;
		}
		this.bytes[track][this.byt] = value;
		nextByte();
	}

	private boolean nextByte()
	{
		++this.byt;
		final boolean nextTrack = this.byt >= BYTES_PER_TRACK;
		if (nextTrack)
		{
			this.byt = 0;
		}
		return nextTrack;
	}

	public boolean isWriteProtected()
	{
		synchronized (this.writable)
		{
			return !this.writable.get();
		}
	}
}

File: epple2sdl/src/diskbytes.cpp
/*
    epple2
    Copyright (C) 2008 by Chris Mosher <chris@mosher.mine.nu>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY, without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <http://www.gnu.org/licenses/>.
*/
#include "diskbytes.h"

#include <vector>
#include <algorithm>
#include <istream>
#include <ostream>
#include <fstream>

DiskBytes::DiskBytes(/*TODO HyperMode& fhyper*/)
{
	unload();
}

DiskBytes::~DiskBytes()
{
}

void DiskBytes::load(const std::string& filePath)
{
	if (isLoaded())
	{
		unload();
	}

	std::ifstream in(filePath.c_str(),std::ios::binary);
	for (int t(0); t < TRACKS_PER_DISK; ++t)
	{
		this->bytes[t].resize(BYTES_PER_TRACK);
		in.read((char*)&this->bytes[t][0],BYTES_PER_TRACK);
	}
	// TODO check file length on all bytes >= 0x96
	// TODO check I/O errors
	in.close();

	this->filePath = filePath;

	std::ofstream outf(filePath.c_str(),std::ios::binary|std::ios::app);
	this->writable = outf.is_open();
	outf.close();

	this->loaded = true;
	this->modified = false;
}

void DiskBytes::save()
{
	if (isWriteProtected() || !isLoaded())
	{
		return;
	}
	std::ofstream out(filePath.c_str(),std::ios::binary);
	for (int t(0); t < TRACKS_PER_DISK; ++t)
		out.write((char*)&this->bytes[t][0],BYTES_PER_TRACK);
	out.flush();
	out.close();

	this->modified = false;
}

void DiskBytes::unload()
{
	this->byt = 0;
	this->writable = true;
	this->loaded = false;
	this->filePath = "";
	this->modified = false;
}

unsigned char DiskBytes::get(const int track)
{
//	if (track < 0 || Drive.TRACKS_PER_DISK <= track)
//	{
//		throw new IllegalStateException();
//	}
	if (!isLoaded())
	{
// TODO		waitIfTooFast();
//		if (!isLoaded())
//		{
			return 0xFF;
//		}
	}
	const unsigned char ret = this->bytes[track][this->byt];
	nextByte();
	return ret;
}
/*
void waitIfTooFast()
{
	if (this->fhyper.isHyper())
	{
		return;
	}

	++this->waitTimes;
	if (this->waitTimes >= 0x100)
	{
		synchronized (this->loaded)
		{
			try
			{
				this->loaded.wait(50);
			}
			catch (InterruptedException e)
			{
				Thread.currentThread().interrupt();
			}
		}
		this->waitTimes = 0;
	}
}
*/
void DiskBytes::put(const unsigned char track, const unsigned char value)
{
	if (TRACKS_PER_DISK <= track)
	{
		throw 0;
	}
	if (isWriteProtected() || !isLoaded())
	{
		return;
	}
	this->bytes[track][this->byt] = value;
	this->modified = true;
	nextByte();
}

void inline DiskBytes::nextByte()
{
	// emulates circular disk track
	++this->byt;
	if (this->byt >= BYTES_PER_TRACK)
	{
		this->byt = 0;
	}
}

